duo.js

```
@@ -4,22 +4,26 @@ var DUO_PREFIX = 'TX'
 var APP_PREFIX = 'APP'
 var AUTH_PREFIX = 'AUTH'
 
+var INIT_EXPIRE = 300
 var DUO_EXPIRE = 300
 var APP_EXPIRE = 3600
 var IKEY_LEN = 20
 var SKEY_LEN = 40
 var AKEY_LEN = 40
 
 /* Exception Messages */
-var ERR_USER = 'ERR|The username passed to sign_request() is invalid.'
-var ERR_IKEY = 'ERR|The Duo integration key passed to sign_request() is invalid'
-var ERR_SKEY = 'ERR|The Duo secret key passed to sign_request() is invalid'
-var ERR_AKEY = 'ERR|The application secret key passed to sign_request() must be at least ' + String(AKEY_LEN) + ' characters.'
-
-exports.ERR_USER = ERR_USER
-exports.ERR_IKEY = ERR_IKEY
-exports.ERR_SKEY = ERR_SKEY
-exports.ERR_AKEY = ERR_AKEY
+var userError = function (method) {
```

Why are these functions now? Shouldn't the caller know that the username passed to `$method` was invalid if the error message is returned from `$method`?

---

The functions just mitigates duplicating error messages. Instead of having 4 error messages for sign_request and 4 error messages for sign_app_blob with just one word different between the respective error messages, I have 4 functions which interpolate the error message with a string depending on which function name you call it with.

---

Sure, but why do we need to include the function name in the message? I.e. if I call `module.function_call(...)` and get `ERR|...` from that function, shouldn't I know that it occurred in `function_call`?

Realistically, this should probably throw an exception instead, which would include all the traceback information. But I don't know if we want to make that change yet.